v3/mathjax3-ts/core/MathItem.ts

```diff
@@ -0,0 +1,331 @@
+ */
+
+export interface ProtoItem {
+    math: string;      // The math expression itself
+    start: Location;   // The starting location of the math
+    end: Location;     // The ending location of the math
+    open?: string;     // The opening delimiter
+    close?: string;    // The closing delimiter
+    n?: number;        // The index of the string in which this math is found
+    display: boolean;  // True means display mode, false is inline mode
+}
+
+/*****************************************************************/
+/*
+ *  Implements the MathItem class
+ */
+
+export abstract class AbstractMathItem implements MathItem {
+
+    public static STATE = {
+        UNPROCESSED: 0,
+        COMPILED: 1,
+        TYPESET: 2,
+        INSERTED: 3
+    };
+
+    public math: string;
+    public inputJax: InputJax;
+    public display: boolean;
+    public start: Location;
+    public end: Location;
+    public root: MmlNode = null;
+    public typesetRoot: Element = null;
+    protected _state: number = STATE.UNPROCESSED;
+    public metrics: Metrics = {} as Metrics;
+    public bbox: BBox = {};
+    public inputData: OptionList = {};
+    public outputData: OptionList = {};
+
+    /*
+     * @param{string} math      The math expression for this item
+     * @param{Inputjax} jax     The input jax to use for this item
+     * @param{boolean} display  True if display mode, false if inline
+     * @param{Location} start   The starting position of the math in the document
+     * @param{Location} end     The ending position of the math in the document
+     * @constructor
+     */
+    constructor (math: string, jax: InputJax, display: boolean = true,
+                 start: Location = {i: 0, n: 0, delim: ''},
+                 end: Location = {i: 0, n: 0, delim: ''}) {
+        this.math = math;
+        this.inputJax = jax;
+        this.display = display;
+        this.start = start;
+        this.end = end;
+        this.root = null;
+        this.typesetRoot = null;
+        this.metrics = {} as Metrics;
+        this.bbox = {};
+        this.inputData = {};
+        this.outputData = {};
+    }
+
+    /*
+     * @override
+     */
+    public compile(document: MathDocument) {
+        if (this.state() < STATE.COMPILED) {
+            this.root = this.inputJax.compile(this);
+            this.state(STATE.COMPILED);
+        }
+    }
+
+    /*
+     * @override
+     */
+    public typeset(document: MathDocument) {
+        if (this.state() < STATE.TYPESET) {
+            if (this.display === null) {
+                this.typesetRoot = document.outputJax.escaped(this, document);
+            } else {
+                this.typesetRoot = document.outputJax.typeset(this, document);
+            }
+            this.state(STATE.TYPESET);
+        }
+    }
+
+    /*
+     * @override
+     */
+    public addEventHandlers() {}
+
+    /*
+     * @override
+     */
+    public updateDocument(document: MathDocument) {}
```

Comment:
No, just like `addEventHandlers()`, this method does nothing if you don't implement it.  You don't *have* to implement it.